src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,350 @@
+                        {
+                            // extension is not in the _extensions dictionary anymore, so it's AssemblyLoadContext must be unloaded
+                            mustCleanupExtension = true;
+                            throw new InvalidOperationException($"{assemblyPath} was unloaded while loading handlers.");
+                        }
+
+                        var duplicateHandler = _handlers.Keys.Intersect(messageHandlers.Keys).Concat(
+                            _documentHandlers.Keys.Intersect(messageHandlers.Keys)).Concat(
+                            _handlers.Keys.Intersect(messageDocumentHandlers.Keys)).Concat(
+                            _documentHandlers.Keys.Intersect(messageDocumentHandlers.Keys)).FirstOrDefault();
+
+                        if (duplicateHandler is not null)
+                        {
+                            throw new InvalidOperationException($"Handler name {duplicateHandler} is already registered.");
+                        }
+
+                        foreach (var handler in messageHandlers)
+                        {
+                            _handlers.Add(handler.Key, handler.Value);
+                        }
+
+                        foreach (var handler in messageDocumentHandlers)
+                        {
+                            _documentHandlers.Add(handler.Key, handler.Value);
+                        }
+
+                        extension.Assemblies[assemblyFileName] = new()
+                        {
+                            Handlers = messageHandlers.Keys.ToHashSet(),
+                            DocumentHandlers = messageDocumentHandlers.Keys.ToHashSet(),
+                        };
+
+                        return ValueTask.FromResult<RegisterHandlersResponse>(
+                            new(messageHandlers.Keys.ToArray(), messageDocumentHandlers.Keys.ToArray()));
+                    }
+                }
+                catch (Exception e)
+                {
+                    Logger.Log(
+                        FunctionId.CustomMessageHandlerService_HandleCustomMessageAsync,
+                        $"Error loading handlers from {assemblyPath}: {e}",
+                        LogLevel.Error);
+
+                    if (mustCleanupExtension)
+                    {
+                        extension.AssemblyLoadContext.Unload();
+                    }
+                    else
+                    {
+                        extension.Assemblies[assemblyFileName] = null;
+                    }
+                    throw;
+                }
+            }
+        }
+
+        Assembly? ResolveExtensionAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
+        {
+            try
+            {
+                return _defaultLoadContext.LoadFromAssemblyName(assemblyName);
+            }
+            catch
```

Comment:
this seems problematic.

Reply:
@dibarbet, I think you suggested this approach. So I would like your input as well before I make changes.